#2 - Fix user can submit 10-digit amount and gives an unexpected error

bernhardoj opened the pull request:
### Details
When we select a currency without decimals, we can input up to 10 digit of amount which leads to a BE error. This PR fix it.


### Fixed Issues
$ https://github.com/Expensify/App/issues/42084
$ https://github.com/Expensify/App/issues/42780
PROPOSAL: https://github.com/Expensify/App/issues/42084#issuecomment-2109204723


### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->
Same as QA Steps
- [x] Verify that no errors appear in the JS console

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->
Same as QA Steps
### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->
A.
1. Open a submit expense page
2. Change the currency to IQD
3. Verify you can't enter an amount with more than 8 digit
4. Select any participant and submit the expense
5. Verify the request is successful and won't give an error

B.
1. Go to workspace settings
2. Go to Distance rates
3. Click add rate
4. Verify you can enter an amount with 3 digit decimals
5. Input 0.655 and save the rate
6. Click the new rate you just created
7. Click Rate to open the rate amount edit page
8. Verify the amount is shows 0.655

- [x] Verify that no errors appear in the JS console

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/50919443/f06fac69-a5bc-44aa-91b3-778cd1930736


</details>

<details>
<summary>Android: mWeb Chrome</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/50919443/1eba1245-e0fa-4a0f-a478-cfeb2af60362


</details>

<details>
<summary>iOS: Native</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/50919443/d8d0c0cb-5e79-41d8-8dfe-87a3c9d9beae


</details>

<details>
<summary>iOS: mWeb Safari</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/50919443/2f69bc1d-828d-4245-a46d-4a95f27e098a


</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/50919443/b6801a9d-23b4-427d-8110-6fdc6c278191


</details>

<details>
<summary>MacOS: Desktop</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/50919443/0ad96676-c4be-41d0-bd38-90fe23689763


</details>

bernhardoj committed:
always convert the amount to cent

bernhardoj committed:
don't allow NaN, Infinity, and number with scientific notation

bernhardoj commented:
Additional recording showing the distance rate with 3 decimals (no trailing zero) is shown

https://github.com/Expensify/App/assets/50919443/06b525bb-2305-48e0-8e47-160d72b83486

ikevin127 commented:
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/56457735/e64b74e3-4e5d-4292-b0a8-245e23e39a73

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/56457735/4bcfe69f-b15a-4fde-8fd8-a4e40873c311

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/56457735/41662214-4f91-4382-b4d2-ae2db92a5e47

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/56457735/2aa6557a-b28e-485c-8dad-59995c472ad0

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/56457735/a4b5f5bf-32be-4ca7-989e-d88dd6c2422a

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/56457735/022f7c1d-94f8-45cd-b2e3-cf10da8826b2

</details>

bernhardoj commented:
I have one question, currently, the max digit is 10.

In staging (or `main`), we can input max. 7 digits + 3 decimals for the distance rate. The max. digit before decimal (.) is 7.

But in this PR, we can input 8 digits + 2 decimals or 7 digits + 3 decimals. The max. digit before decimal (.) is 8.

But we can't input 8 digits + 3 decimals because the max digit is 10, do we want to increase the max digit to 11 for the distance rate?

I'm asking this because when I set the distance rate to `12345678.12`, the BE returns `12345678.120`, with a trailing zero, resulting in 11 digits.

bernhardoj committed:
remove dot when calculating the length

ikevin127 commented:
@marcochavezf @francoisl Any input on the [comment](https://github.com/Expensify/App/pull/42815#issuecomment-2138605623) above ?

bernhardoj commented on src/libs/MoneyRequestUtils.ts:
```diff
@@ -40,16 +47,17 @@ function addLeadingZero(amount: string): string {
     }
 
-    return leadingZeroesLength + (absAmount === '0' ? 2 : absAmount.length);
+    return leadingZeroesLength + (absAmount === '0' ? 2 : stripDotFromAmount(absAmount).length);
```
Also, another way to fix this is by passing the removed-dot-amount to the old regex

```
const amountWithoutDecimalSeparator = stripDotFromAmount(absAmount);
if (/\D/.test(amountWithoutDecimalSeparator)) {
    return CONST.IOU.AMOUNT_MAX_LENGTH + 1;
}
```

Let me know if we want to use this instead.

francoisl replied:
I think I like the current version better, but can we also add unit tests to prevent future regressions please?

bernhardoj replied:
Before adding a test, what do you think about my suggestion [here](https://github.com/Expensify/App/pull/42815#issuecomment-2138678006). Do you think we should do it or keep it as it is for now? 

If we want to do it, then we don't need the test for this specific function.

ikevin127 commented:
@bernhardoj One of the reported deploy-blocker issues (https://github.com/Expensify/App/issues/42780) is still happening on this PR:

<img width="1130" alt="Screenshot 2024-05-29 at 20 35 35" src="https://github.com/Expensify/App/assets/56457735/098804c8-70b3-4d45-a3b9-60864ba6d7a1">

bernhardoj commented:
It's happening on `main` too. That's weird because IQD doesn't have any decimal. We can't input 0.7 in submit expense amount page but we can in distance rate amount.

bernhardoj commented:
I found the root cause of the above issue, but I think it should be fixed on a separate issue.

We pass an extra 1 decimal to every currency
https://github.com/Expensify/App/blob/f6aa6c20a18e2d64f3143815f5266d064ce7cb6a/src/pages/workspace/distanceRates/PolicyDistanceRateEditPage.tsx#L94

Then, we validate the amount, however, the BE returns the decimal with trailing zeros (0.700 for example)
https://github.com/Expensify/App/blob/f6aa6c20a18e2d64f3143815f5266d064ce7cb6a/src/components/AmountForm.tsx#L129-L134

so it's not valid because the extra decimal is 1, so the decimals is stripped.

ikevin127 commented:
Interesting, I thought it's coming from our first PR somehow because the testers got back with this  https://github.com/Expensify/App/issues/42780#issuecomment-2138162745 in which the issue doesn't seem to happen once the PR was reverted.

But I guess it's related to the `IQD` currency which might be different from `MYR` ? 🤔

bernhardoj commented:
Yep, it depends with the currency because IQD doesn't support decimals.

Btw, regarding my question [here](https://github.com/Expensify/App/pull/42815#issuecomment-2138605623), I just realized that if we increase the max digit to 11, we can then input 9 digits + 2 decimals which is invalid. So, we can't do that.

I propose that we refactor `validateAmount` a bit (maybe on another issue too?).
https://github.com/Expensify/App/blob/9e9d8cdd97f3d3d67fa4be78e89bbd4a29aa1702/src/libs/MoneyRequestUtils.ts#L58-L65

The refactor will fully rely on the regex. So, we will remove `calculateAmountLength`. `calculateAmountLength` isn't powerful enough because it just count the digit length. We will update the `regexString` to validate the max digit of the amount before the decimal and the decimal max digit.

Currently, the regex matches this number: 123,123.456 but I don't see any usage in our app that allows the thousand separator(?) in the amount. So, we can update the regex to:
```
const regexString =
    decimals === 0
        ? `^\\d{1,${amountMaxLength}}$`
        : `^\\d{1,${amountMaxLength}}(\\.\\d{0,${decimals}})?$`;
```

amountMaxLength will be the max length before the decimal. Currently, amountMaxLength is the max length of all digits including the decimals. So for example, for distance rate, amountMaxLength will be 8 and decimals will be 3.

For tax rate, the max. value is 100%, so we can set the amountMaxLength to 3. This way, we can flexibly control the max. digit for different use cases.

Let me know what you think!

ikevin127 commented:
Please let me know besides the main issue (https://github.com/Expensify/App/issues/42084) this PR fixes, if we're going to cover the other issues as well:
- https://github.com/Expensify/App/issues/42778
- https://github.com/Expensify/App/issues/42780

And if so I think it would be a good idea to update the PR description to include the other issues and **Tests** for each.
Otherwise, let me know what additional changes are going to be part of this PR besides the main issue (https://github.com/Expensify/App/issues/42084).

bernhardoj commented:
@francoisl oh my bad, turns out it's the FE that formatted the amount to have 3 decimals points.
https://github.com/Expensify/App/blob/5699ae382505ed0d962e6170320a367e41e7858c/src/pages/workspace/distanceRates/PolicyDistanceRatesPage.tsx#L105-L107
https://github.com/Expensify/App/blob/5699ae382505ed0d962e6170320a367e41e7858c/src/libs/CurrencyUtils.ts#L118-L123

francoisl commented:
Ok yeah I think I like the idea in [this comment](https://github.com/Expensify/App/pull/42815#issuecomment-2138678006) to rely on a regex to validate the max number of digits and decimals.
Can we do it in this PR? It doesn't seem like a huge change in the end.

bernhardoj commented:
I can do it. I will update it on Monday.

bernhardoj committed:
refactor max amount digit logic

bernhardoj committed:
add unit test

bernhardoj committed:
remove unused function

bernhardoj committed:
prettier

bernhardoj commented:
Updated `MoneyRequestUtils.validateAmount`. Here is the list of the usages:

MoneyRequestConfirmationList (distance amount, 8 digits + 2 decimals):

https://github.com/Expensify/App/assets/50919443/e8a06833-6def-4fe1-8f15-634ce9c019be


AmountForm
    - AmountSelectorModal/AmountPicker/WorkspaceCreateTaxPage (create tax rate page, 4 digits + 4 decimals):

https://github.com/Expensify/App/assets/50919443/aa45ab6f-07b4-43e8-bfc8-e846346dc6ba

    - ValuePage (edit tax rate page, 4 digits + 4 decimals):
   
https://github.com/Expensify/App/assets/50919443/6a20fc3b-53cf-42c3-926e-52e44f72f3a4

    - PolicyDistanceRateEditPage (edit distance rate page 8 digits + 3 decimals):

https://github.com/Expensify/App/assets/50919443/34e22a59-499c-46c8-9262-39f84dc7b65e

    - CreateDistanceRatePage (create distance rate page, 8 digits + 3 decimals):

https://github.com/Expensify/App/assets/50919443/40875f08-133b-402c-ae26-611b5bbffe9c

    - RatePage (no usage)
MoneyRequestAmountInput
    - MoneyRequestAmountForm
    ---- IOURequestStepAmount (IOU amount page, 8 digits + 2 decimals):

https://github.com/Expensify/App/assets/50919443/c8af9814-3e71-4116-b0c4-394c73740ca1

    ---- IOURequestStepTaxAmountPage (IOU tax amount page, 8 digits + 2 decimals)

https://github.com/Expensify/App/assets/50919443/82491a28-3dc6-4c0d-867b-c7039d7d4188

    - MoneyRequestConfirmationList (individual split amount, 8 digits + 2 decimals):

https://github.com/Expensify/App/assets/50919443/659c2b1b-2062-47c5-a17d-58031af4b3e3

individual split amount input is the only one that formats the amount to have `,` separator, but when we edit it, the `,` will be removed or replaced with `.`.
https://github.com/Expensify/App/blob/f701c3e0302b3e1eadf7767433277fea9eac0d6e/src/components/MoneyRequestAmountInput.tsx#L147-L150

Added a unit test too.

bernhardoj committed:
use the correct const

ikevin127 commented:
@francoisl Are you good with the latest changes and behaviour from https://github.com/Expensify/App/pull/42815#issuecomment-2144231871 ? Asking in order to confirm whether I should start working on the reviewer checklist.

##

@bernhardoj Follow-up regarding my  https://github.com/Expensify/App/pull/42815#issuecomment-2140887387, besides the main issue attached in **Fixed issues**, assuming this PR's changes are addressing the other 2 issues, could you add reviewer / QA tests for those too ? Asking in order know what to test when I'm gonna start working on the reviewer checklist.

bernhardoj commented:
@ikevin127 updated the test step.

ikevin127 commented:
I'll have the reviewer checklist ready by end of day today. @bernhardoj When you get a chance, please solve the conflicts.

francoisl commented:
> @francoisl Are you good with the latest changes and behaviour from https://github.com/Expensify/App/pull/42815#issuecomment-2144231871 ? Asking in order to confirm whether I should start working on the reviewer checklist.

Woops sorry missed that comment – yes the changes in the videos look good to me.

ikevin127 commented:
@bernhardoj I tested both scenarios and found some issues with scenario B, only with the `IQD` currency:
1. Create WS and set default currency to `IQD`, then enable **Distance rates** and try to Edit rate (the default rate) -> edit field will show 0 (issue [#42780](https://github.com/Expensify/App/issues/42780), but for `IQD` only).
^ Additional observation: `IQD` currency allows 8 digits but only 1 decimal.

I think this is the last issue that needs to be addressed before I can ✅ **Approve** for final review.

bernhardoj committed:
Merge branch 'main' into fix/42084-limi-amount-to-8-digit

bernhardoj commented:
It's the same as [this one](https://github.com/Expensify/App/pull/42815#issuecomment-2138614948), right?

<img width="179" alt="image" src="https://github.com/Expensify/App/assets/50919443/82b32715-94ac-437d-84be-3ffe1e2f6d0e">

ikevin127 commented:
Yes, but it only happens for `IQD` (and probably other similar currencies) and only when trying to **edit** the existing distance rates -> it's simply showing 0 instead of the rate with decimals.

bernhardoj commented:
Yeah, it happens with any currency that doesn't support decimals, such as yen. I thought we want to fix it on another issue based on this [comment](https://github.com/Expensify/App/pull/42815#issuecomment-2138629692) haha.

I'm not sure though what would be the best way to handle that. The currency doesn't support decimal, but in FE, we always add 1 extra decimal support, and BE accepts it too. Even the default distance rate has 2 decimals (0.67) for non-decimal (IQD) currency.

I think it needs to be discussed on another issue or if we want a temporary fix, we can set the distance rate decimal to always be 3, instead of currency decimal + 1 extra decimal.

cc: @francoisl

ikevin127 reviewed (approved):
Given the above context (https://github.com/Expensify/App/pull/42815#issuecomment-2154126801) and the fact that the only issue left was/is already present on staging and was decided on as out of scope [here](https://github.com/Expensify/App/pull/42815#issuecomment-2138629692), I'm approving with the mention that if the out of scope issue will surface post-merge, it will not be considered a regression of this PR.

marcochavezf reviewed (approved):
LGTM, I agree we should handle the edge case for currencies without decimal support as out-of-scope for now

@francoisl all yours

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/9474095450) to production by https://github.com/luacmartins in version: 1.4.81-11 🚀

platform | result
---|---
🤖 android 🤖|failure ❌
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅
